### DISCUSSION TITLE
Connecting iPhone with USB

### DISCUSSION THREAD

**Discussion by TheProgramableTurtle** [microsoft/WSL#5525] [General]
**Is your feature request related to a problem? Please describe.**
I want to connect my iPhone to communicate with it using command-line tools in WSL but am unable to get those tools to detect the device. This is the same as #5169 but my problem was not resolved after trying everything suggested.

**Describe the solution you'd like**
Optimally, proper USB passthrough, but [this guide](https://github.com/rpasek/usbip-wsl2-instructions) and the solutions presented in #5169 do make the USB device show with `lsusb`. However, command-line tools like [libimobiledevice](https://libimobiledevice.org/) are unable to detect the device with `idevicepair list`.

**Describe alternatives you've considered**
Using the methods above I have built the WSL kernel with USB functionality enabled and also used the kernel provided by the person in the other issue that solved the problem for that person, used the Virtual Here USB server on my Windows host and the Virtual Here GUI Linux client (using VcXsrv as the XServer if that makes a difference), and several different USB ports with no success.

**Additional context**
I have tried with 5 devices (and iPhone 4, a 5S, two 6Ss, and an HTC Wildfire). For the Android, I tested the connection using gmtp but it was not detected. None of the iDevices were detected in normal mode, but the 5S (the only phone I tested this part with) was not detected in recovery mode either but was in DFU mode.
I made this issue as a feature request because USB isn't officially supported yet (and according to Microsoft, won't be for a very long time) and so this isn't really a bug, but if there is a temporary solution or alternative that would be great. (I do have access to a Linux machine so it isn't the end of the world, I'm just trying to see if it's possible).

**Comment by TheProgramableTurtle** [comment]
[listmod.txt](https://github.com/microsoft/WSL/files/4869604/listmod.txt)
Just realised that when running `lsmod` on WSL it doesn't show any modules and I can't modprobe anything, going to try rebuilding the kernel. 
Edit: Same across all my distros with a bunch of different kernels.
Edit again: When building the kernel, `sudo make modules_install` doesn't show any modules, just 
`DEPMOD  4.19.104-microsoft-standard`
Edit 3: Reset changes, copied over .config with drivers enabled, built, still no modules
```
Building modules, stage 2.
MODPOST 0 modules
```
Edit 4: Looking at the .config file it seems USBIP support was disabled, now modules show up.

**Comment by TheProgramableTurtle** [comment]
Ok I'm very confused, I deleted the entire WSL repo and cloned it again to start afresh since `git reset` didn't work, but, despite there not being a .config file, `sudo make menuconfig` still shows all the options from last time. When I build, no modules are built and `lsmod` is blank when using the kernel. Does it have some ever-lasting memory or something, and why when the .config has modules selected are those modules not built (building with modules selected as "m")? Think I might just delete the entire distro and start again.
Edit: Ok I'm stupid, apparently it takes the current configuration from /boot. But still, why don't the modules build?

**Comment by TheProgramableTurtle** [comment]
Using your kernel it seems `ipheth` is loaded
`usbcore: registered new interface driver ipheth`
Yeah, I thought that was how it worked, but then are modules with (m) supposed to show in `lsmod`?

**Comment by TheProgramableTurtle** [comment]
dmesg shows
```
[  883.605753] vhci_hcd vhci_hcd.0: pdev(0) rhport(0) sockfd(9)
[  883.605755] vhci_hcd vhci_hcd.0: devid(4) speed(3) speed_str(high-speed)
[  883.952848] usb 1-1: new high-speed USB device number 3 using vhci_hcd
[  884.102839] usb 1-1: SetAddress Request (3) to port 0
[  884.190220] usb 1-1: New USB device found, idVendor=05ac, idProduct=1297, bcdDevice= 3.20
[  884.190222] usb 1-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[  884.190223] usb 1-1: Product: iPhone
[  884.190224] usb 1-1: Manufacturer: Apple Inc.
[  884.190225] usb 1-1: SerialNumber: f29e22a4907ddf8df8b7641911502d10291c5a69
```
and the VirtualHere logs show
```
2020-07-03 15:28:07 INFO :Found High speed device [05ac:1297] "Apple Inc., iPhone" at address 11
2020-07-03 15:28:14 INFO :Unmanaging device 11 [05ac:1297]
2020-07-03 15:28:14 INFO :Found High speed device [05ac:1297] "Apple Inc., iPhone" at address 11
2020-07-03 15:28:14 INFO :Device 11 [05ac:1297] BOUND to connection 1
```

**Comment by TheProgramableTurtle** [comment]
Using the Ubuntu 18.04 build from https://cloud-images.ubuntu.com/bionic/current/ (I have Ubuntu 20.04 from the Microsoft Store but I've got that set up for doing a bunch of other stuff) and Ubuntu 18.04 in the Linux machine. I'll try using Microsoft Store Ubuntu and see if it makes a difference.

**Comment by TheProgramableTurtle** [comment]
Nope, same result, `lsusb` shows the device but `idevicepair list` does not.

**Comment by TheProgramableTurtle** [comment]
Compiled a kernel with those options enabled but still doesn't work. There's nothing about HFS or HFSPLUS or Apple filesystems in `dmesg` though.

**Comment by TheProgramableTurtle** [comment]
I've tried a bunch of different USB ports, both 2.0 and 3.0, as well as ones on the case and mobo, and also tried with a jailbroken device, but still no luck. Going to try with a more recent device and see if that makes a difference. Though the fact that adb couldn't find the android either makes me think it might be a hardware problem...

**Comment by TheProgramableTurtle** [comment]
Nope, nothing, but for some reason, an iPhone 8 shows as Apple, Inc. iPhone5/5C/5S/6.
Actually, they all do, except for the iPhone 4 which shows as an Apple, Inc. iPhone 4.

**Comment by TheProgramableTurtle** [comment]
After removing the default Windows drivers and manually installing the Apple drivers, as suggested [here](https://github.com/cezanne/usbip-win/issues/149#issuecomment-622367314), I got the stable version of the VirtualHere server (previously I had been using the beta) working with connecting the iPhone 4, but the device is still not detected by CLI tools. When running `usbip port` from WSL (I installed USBIP on WSL) I get this,
```
usbip: error: failed to open /usr/share/hwdata//usb.ids
Imported USB devices
====================
libusbip: error: fopen
libusbip: error: read_record
Port 00: <Port in Use> at High Speed(480Mbps)
       unknown vendor : unknown product (05ac:1297)
       1-1 -> unknown host, remote port and remote busid
           -> remote bus/dev 000/004
```
which, ignoring the errors for now, shows that programs in WSL are able to detect the device but are unable to identify it as an iPhone. I know this device is the iPhone because `lsusb` shows
```
Bus 001 Device 004: ID 05ac:1297 Apple, Inc. iPhone 4
```
From what I can tell, the device can't be identified because there is no driver for it installed (or running).

**Comment by TheProgramableTurtle** [comment]
Ok nvm I broke it again :/ I think the stub drivers from USBIP are what was making it work, but with them installed every time I connected a USB device it was taken by the driver and I had to manually change it (which is difficult without a mouse). Even if this could work it would be ridiculously difficult. With the iPhone 4 in DFU mode (the mode checkra1n was able to detect it) `usbip port` shows
```
usbip: error: failed to open /usr/share/hwdata//usb.ids
Imported USB devices
====================
libusbip: error: fopen
libusbip: error: read_record
Port 00: <Port in Use> at High Speed(480Mbps)
       unknown vendor : unknown product (05ac:1227)
       1-1 -> unknown host, remote port and remote busid
           -> remote bus/dev 000/005
```
Same errors as last time, same ambiguity as last time, so I don't know why the device can be detected in DFU mode but not normal. Worth noting that checkra1n is the only tool that I could get to detect it in DFU mode, `idevicepair list` shows nothing. Just to confirm, `lsusb` shows
`Bus 001 Device 003: ID 05ac:1227 Apple, Inc. Mobile Device (DFU Mode)`
as I would expect.
Also, whenever I force restart from DFU mode on my iPhone 4 or 5S the screen flickers like crazy. Or maybe I'm going crazy...

**Comment by TheProgramableTurtle** [comment]
Seems obvious, but usbip can't identify the device because it has nothing to reference. Downloading [usb.ids](http://www.linux-usb.org/usb.ids) and placing it in the required folder fixed this, now I get
```
Imported USB devices
====================
libusbip: error: fopen
libusbip: error: read_record
Port 00: <Port in Use> at High Speed(480Mbps)
       Apple, Inc. : iPhone 4 (05ac:1297)
       1-1 -> unknown host, remote port and remote busid
           -> remote bus/dev 000/004
```
But nothing else can identify it, going to check how libimobiledevice does it (can't with checkra1n since it's closed source).

**Comment by therealkenc** [comment]
USB device support is open WSL #2195. 

This submission is a discussion regarding [rpasek/usbip-wsl2-instructions](https://github.com/rpasek/usbip-wsl2-instructions). That repository is the place I'd hang out for questions about usage, what works, what does does not, etc etc (if it were me). But do feel encouraged to continue the discussion in the discussion forum.
